### PR TITLE
Use wget for http/https package in downloader container

### LINKED ISSUE #545: Deploying functions without privileged access
With standard deployments via Pulsar API we needed to have admin tokens as a basis for running functions on our JWT secured clusters. We hoped that by moving to function mesh deployments, privileged access to pulsar is not needed anymore.
### Actual behaviour (v0.8):
Function pod does not start download (without privileged access) even the function is downloaded from an external HTTPS location. 
```shell
HTTP 401 Unauthorized
Reason: HTTP 401 Unauthorized
```
If an admin token is configured function downloads and runs just fine.
```shell
Downloaded successfully
shardId=0
[2022-12-09 14:18:55 +0000] [INFO] aa python_instance_main.py: Starting Python instance with Namespace(client_auth_params='file:///var/run/secrets/pulsar/auth-token/token', client_auth_plugin='org.apache.pulsar.client.impl.auth.AuthenticationToken', cluster_name='kubernetes', dependency_repository=None, expected_healthcheck_interval=-1, extra_dependency_repository=None, function_details='{"tenant":"public","namespace":"default","name":"exclamation-mark","className":"exclamationMark.ExclamationFunction","autoAck":true,"parallelism":1,"source":{"typeClassName":"[B","inputSpecs":{"persistent://public/default/py-function-input-topic":{}},"subscriptionPosition":"EARLIEST"},"sink":{"typeClassName":"[B","topic":"persistent://public/default/py-function-output-topic","forwardSourceMessageProperty":true},"resources":{"cpu":1,"ram":"1100000000"},"componentType":"FUNCTION"}', function_id='0-bedb7185-3b4f-4b7e-8a3a-bdfa02c510e7', function_version='0', hostname_verification_enabled=None, install_usercode_dependencies=Tr...
2022-12-09 14:18:55.694 INFO [140709881481024] Client:88 | Subscribing on Topic :persistent://public/default/py-function-input-topic
2022-12-09 14:18:55.694 INFO [140709881481024] ClientConnection:182 | [<none> -> pulsar+ssl://prod01-neuron-pulsar-proxy.prod01-neuron-pulsar.svc.cluster.local:6651] Create ClientConnection, timeout=10000
2022-12-09 14:18:55.779 INFO [140709881481024] ConnectionPool:96 | Created connection for pulsar+ssl://prod01-neuron-pulsar-proxy.prod01-neuron-pulsar.svc.cluster.local:6651
2022-12-09 14:18:55.793 INFO [140709818574592] ClientConnection:368 | [10.130.4.14:53720 -> 172.30.72.61:6651] Connected to broker
2022-12-09 14:18:55.814 INFO [140709818574592] HandlerBase:64 | [persistent://public/default/py-function-input-topic, public/default/exclamation-mark, 0] Getting connection from pool
2022-12-09 14:18:55.815 INFO [140709818574592] ClientConnection:182 | [<none> -> pulsar+ssl://prod01-neuron-pulsar-proxy.prod01-neuron-pulsar.svc.cluster.local:6651] Create ClientConnection, timeout=10000
2022-12-09 14:18:55.820 INFO [140709818574592] ConnectionPool:96 | Created connection for pulsar+ssl://prod01-neuron-pulsar-broker-2.prod01-neuron-pulsar-broker.prod01-neuron-pulsar.svc.cluster.local:6651
2022-12-09 14:18:55.824 INFO [140709818574592] ClientConnection:370 | [10.130.4.14:53728 -> 172.30.72.61:6651] Connected to broker through proxy. Logical broker: pulsar+ssl://prod01-neuron-pulsar-broker-2.prod01-neuron-pulsar-broker.prod01-neuron-pulsar.svc.cluster.local:6651
2022-12-09 14:18:55.842 INFO [140709818574592] ConsumerImpl:224 | [persistent://public/default/py-function-input-topic, public/default/exclamation-mark, 0] Created consumer on broker [10.130.4.14:53728 -> 172.30.72.61:6651]
[2022-12-09 14:18:55 +0000] [INFO] aa server.py: Serving InstanceCommunication on port 9093
```
On enterprise level this prohibits us to run independent use cases on a shared Pulsar instance.

### Desired behaviour:
Functions should be deployable with minimum privileges and therefore use just the role permissions on the topics they consume / produce.

### Function CR
```yaml
apiVersion: compute.functionmesh.io/v1alpha1
kind: Function
metadata:
  name: exclamation-mark
spec:
  tenant: public #tenant of Pulsar Function (default is "public")
  namespace: default #namespace of Pulsar Function (default is "default")
  python:
    pyLocation: https://nexus.ourrepo.corp/nexus/repository/1st-app-rel/neuron/application-neuron-playground/function/exclamationMark/67a3a0518462ac397ab33c1d68d9daf3c08448e3/exclamationMark-67a3a0518462ac397ab33c1d68d9daf3c08448e3.zip
    py: exclamationMark.zip # downloaded artifact from "pyLocation" will be stored with this filename
  className: exclamationMark.ExclamationFunction
  pulsar:
    pulsarConfig: functionmesh-pulsar-config # functionmesh-pulsar-config configmap reference
    {{ if .Values.idp.clientId }}authSecret: functionmesh-authentication-config{{ end }}
  input:
    topics:
    - persistent://public/default/py-function-input-topic
  output:
    topic: persistent://public/default/py-function-output-topic
  pod:
    serviceAccountName: functions-worker      
  {{- if .Values.idp.clientId }} # secured pulsar
    volumes:
    - name: token-volume
      secret:
        items:
          - key: token
            path: token
        optional: false
        secretName: functionmesh-authentication-secret
  volumeMounts:
  - mountPath: /var/run/secrets/pulsar/auth-token
    name: token-volume
  {{- end }}
  resources:
    requests:
      cpu: "100m"
      memory: 1100M
    limits:
      cpu: "200m" #disabled for debugging
      memory: 1100M
  replicas: 1
```

### PR DESCRIPTION
**Pull request by jiangpengcheng**
Fixes #545

### Motivation

For http/https packages, we don't need to talk with pulsar cluster to download them

### Modifications

Use wget for http/https packages

### Verifying this change

- [x] Make sure that the change passes the CI checks.

- [x] This change added tests and can be verified as follows:

  - *Added integration tests for end-to-end deployment with download package from http*

### Documentation

Check the box below.

Need to update docs? 

- [x] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [ ] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: controllers/spec/common.go [hunk 1]
```diff
@@ -511,6 +516,10 @@ func getLegacyDownloadCommand(downloadPath, componentPackage string, authProvide
 func getDownloadCommand(downloadPath, componentPackage string, tlsProvided, authProvided bool, tlsConfig TLSConfig,
 	authConfig *v1alpha1.AuthConfig) []string {
 	var args []string
+	if hasHTTPPrefix(downloadPath) {
+		args = append(args, "wget", downloadPath, "-O", componentPackage)
```
Comment (freeznet) [inline controllers/spec/common.go:520]: do we need to check the availability of `wget` first, then call it to download?

**Review by freeznet — APPROVED** [review]

File: controllers/spec/common.go (hunk 1 above)
Comment (jiangpengcheng) [reply controllers/spec/common.go:520]: the `pulsarctl` image has `wget`, I think we can update the doc to inform users what a `downloadContainer` requires (currently, it requires pulsarctl and wget)

File: controllers/spec/common.go [hunk 2]
```diff
@@ -511,6 +516,10 @@ func getLegacyDownloadCommand(downloadPath, componentPackage string, authProvide
 func getDownloadCommand(downloadPath, componentPackage string, tlsProvided, authProvided bool, tlsConfig TLSConfig,
 	authConfig *v1alpha1.AuthConfig) []string {
 	var args []string
+	if hasHTTPPrefix(downloadPath) {
```
Comment (freeznet) [inline controllers/spec/common.go:519]: Another question: any chance to make this feature backport to `getLegacyDownloadCommand`?

**Review by freeznet — APPROVED** [review]

**Review by tpiperatgod — APPROVED** [review]

File: controllers/spec/common.go (hunk 2 above)
Comment (jiangpengcheng) [reply controllers/spec/common.go:519]: looks like our runtime image neither have `wget` nor `curl`, so there will be some compatibility issues if add this feature to `getLegacyDownloadCommand`:

- we can add `wget` or `curl` to runtime images and add this feature
- but we also need to deprecate old runtime images or check whether given runtime images have `wget` or `curl` installed

I think there should be a general process for such a procedure(when we need to update runtime images):

- add `wget` or `curl` or other required utils to runtime images and release new versions
- add the new feature to FunctionMesh
- release FunctionMesh vx.y.z, and make an announcement that FunctionMesh vx.y.z is not compatible with runtime images older than va.b.c
- for user's custom runtime images, we list the requirements(what the image should contain)

HDYT? @freeznet @tpiperatgod @nlu90 ?

Comment (tpiperatgod) [reply controllers/spec/common.go:519]: This can be easily solved by Buildpacks, but we currently need a user-friendly image build workflow...

Comment (jiangpengcheng) [reply controllers/spec/common.go:519]: are Buildpacks-related commits merged? or docs updated in functionmesh.io?

Comment (tpiperatgod) [reply controllers/spec/common.go:519]: has provided the docs and demos

**Pull request merged**